File: Samples/Objective-C/SignatureExample/SignatureExample/SignatureIntegrityController.m
//
//  SignatureIntegrityController.m
//  SignatureExample
//
//  Created by Joss Giffard-Burley on 08/08/2014.
//  Copyright (c) 2014 Wacom. All rights reserved.
//

#import "SignatureIntegrityController.h"
#import "UIImage+TintImage.h"
#import <WacomSignatureSDK/WacomSignatureSDK.h>

@interface SignatureIntegrityController ()

@property (strong, nonatomic) IBOutlet UIButton *okButton;
@property (strong, nonatomic) IBOutlet UILabel *who;
@property (strong, nonatomic) IBOutlet UILabel *when;
@property (strong, nonatomic) IBOutlet UILabel *why;
@property (strong, nonatomic) IBOutlet UILabel *captureArea;
@property (strong, nonatomic) IBOutlet UILabel *singatureArea;
@property (strong, nonatomic) IBOutlet UILabel *digitizer;
@property (strong, nonatomic) IBOutlet UILabel *driver;
@property (strong, nonatomic) IBOutlet UILabel *machineOS;
@property (strong, nonatomic) IBOutlet UILabel *hashType;
@property (strong, nonatomic) IBOutlet UILabel *integrityStatus;
@property (strong, nonatomic) IBOutlet UILabel *documentHashType;
@property (strong, nonatomic) IBOutlet UILabel *documentHashStatus;
@property (strong, nonatomic) IBOutlet UILabel *extraItemsLabel;
@property (strong, nonatomic) IBOutlet UIButton *extraItemsButton;
@property (strong, nonatomic) IBOutlet UIView *integView;
@property (strong, nonatomic) UIPopoverController *extraDatapopoverController;

- (SIGNATURE_HASH_TYPE)getHashTypeForSigObj;
- (SIGNATURE_HASH_TYPE)getDocHashTypeForSigObj;
- (NSString *)stringForHashType:(SIGNATURE_HASH_TYPE)hashType;
- (NSString *)validationStringForSignature;
- (NSString *)validationStringForDocumentData;
- (IBAction)extraItemsButtonTapped:(id)sender;

@end

@implementation SignatureIntegrityController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.okButton.layer.cornerRadius = 5.0f;
    [self.extraItemsButton setImage:[[UIImage imageNamed:@"action_about"] tintedImageWithColour:[UIColor colorWithRed:0.5f green:0.5f blue:0.5f alpha:1.0f]] forState:UIControlStateNormal];
}


- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    self.who.text = self.sigObj.who;
    self.why.text = self.sigObj.why;
    self.when.text = [NSDateFormatter localizedStringFromDate:self.sigObj.when dateStyle:NSDateFormatterShortStyle timeStyle:NSDateFormatterShortStyle];
    self.captureArea.text = [NSString stringWithFormat:@"%.0fx%.0f", self.sigObj.captureAreaBounds.width, self.sigObj.captureAreaBounds.height];
    self.singatureArea.text = [NSString stringWithFormat:@"%.0fx%.0f", self.sigObj.signatureAreaBounds.width, self.sigObj.signatureAreaBounds.height];
    self.digitizer.text = self.sigObj.additionalData[kCaptureDigitizer];
    self.driver.text = self.sigObj.additionalData[kCaptureDigitizerDriver];
    self.machineOS.text = self.sigObj.additionalData[kCaptureMachineOS];
    self.hashType.text = [self stringForHashType:[self getHashTypeForSigObj]];
    self.integrityStatus.text = [self validationStringForSignature];
    self.documentHashType.text = [self stringForHashType:[self getDocHashTypeForSigObj]];
    self.documentHashStatus.text = [self validationStringForDocumentData];
    
    if([self.sigObj.extraDataItems count] > 0) {
        self.extraItemsLabel.text = [NSString stringWithFormat:@"%lu Extra Data Items:",  (unsigned long)self.sigObj.extraDataItems.count];
    } else {
        self.extraItemsLabel.hidden = YES;
        self.extraItemsButton.hidden = YES;
    }
}

- (SIGNATURE_HASH_TYPE)getHashTypeForSigObj {
    SIGNATURE_INTEGRITY_STATUS status;

    //MD5
    status = [self.sigObj checkIntegrity:MD5_HASH];
    if(status == INTEGRITY_MISSING) {
        return(NO_HASH);
    } else if((status == INTEGRITY_OK) || (status == INTEGRITY_FAIL) || (status == INTEGRITY_INVALID_KEY)) {//we know the type
        return(MD5_HASH);
    }
    
    //SHA-1
    status = [self.sigObj checkIntegrity:SHA_1_HASH];
    if(status != INTEGRITY_WRONG_TYPE) { //must be OK or FAIL at this point so we have the correct type as missing is accounted for above
        return(SHA_1_HASH);
    }
    
    //SHA-224
    status = [self.sigObj checkIntegrity:SHA_224_HASH];
    if(status != INTEGRITY_WRONG_TYPE) {
        return(SHA_224_HASH);
    }
    
    //SHA-256
    status = [self.sigObj checkIntegrity:SHA_256_HASH];
    if(status != INTEGRITY_WRONG_TYPE) {
        return(SHA_256_HASH);
    }
    
    //SHA-384
    status = [self.sigObj checkIntegrity:SHA_384_HASH];
    if(status != INTEGRITY_WRONG_TYPE) {
        return(SHA_384_HASH);
    }
    
    //SHA-512
    status = [self.sigObj checkIntegrity:SHA_512_HASH];
    if(status != INTEGRITY_WRONG_TYPE) {
        return(SHA_512_HASH);
    }
    
    //MD5-MAC
    status = [self.sigObj checkIntegrity:MD5_MAC_HASH];
    if(status != INTEGRITY_WRONG_TYPE) {
        return(MD5_MAC_HASH);
    }
    
    return(NO_HASH);
}

- (SIGNATURE_HASH_TYPE)getDocHashTypeForSigObj {
    SIGNATURE_SIGNED_DATA_STATUS status;

    status = [self.sigObj checkSignatureData:MD5_HASH withData:self.hashData];
    if(status == DATA_NO_DATA || status == DATA_NO_HASH) {
        return(NO_HASH);
    }
    
    if(status != DATA_BAD_TYPE) {
        return(MD5_HASH);
    }
    
    status = [self.sigObj checkSignatureData:SHA_1_HASH withData:self.hashData];
    if(status != DATA_BAD_TYPE) {
        return(SHA_1_HASH);
    }
    
    status = [self.sigObj checkSignatureData:SHA_224_HASH withData:self.hashData];
    if(status != DATA_BAD_TYPE) {
        return(SHA_224_HASH);
    }
    
    status = [self.sigObj checkSignatureData:SHA_256_HASH withData:self.hashData];
    if(status != DATA_BAD_TYPE) {
        return(SHA_256_HASH);
    }
    
    status = [self.sigObj checkSignatureData:SHA_384_HASH withData:self.hashData];
    if(status != DATA_BAD_TYPE) {
        return(SHA_384_HASH);
    }
    
    status = [self.sigObj checkSignatureData:SHA_512_HASH withData:self.hashData];
    if(status != DATA_BAD_TYPE) {
        return(SHA_512_HASH);
    }
    
    //MD5-MAC
    status = [self.sigObj checkSignatureData:MD5_MAC_HASH withData:self.hashData];
    if(status != DATA_BAD_TYPE) {
        return(MD5_MAC_HASH);
    }
    
    return(NO_HASH);
}

- (NSString *)stringForHashType:(SIGNATURE_HASH_TYPE)hashType {
    switch (hashType) {
        case NO_HASH:
            return(@"No Hash");
        case MD5_HASH:
            return(@"MD5");
        case MD5_MAC_HASH:
            return (@"MD5 with MAC-KEY");
        case SHA_1_HASH:
            return(@"SHA-1");
        case SHA_224_HASH:
            return(@"SHA-224");
        case SHA_256_HASH:
            return(@"SHA-256");
        case SHA_384_HASH:
            return(@"SHA-384");
        case SHA_512_HASH:
            return(@"SHA-512");
    }
}

- (NSString *)validationStringForSignature {
    SIGNATURE_HASH_TYPE hash = [self getHashTypeForSigObj];
    
    if(hash == NO_HASH) {
        return(@"No Hash");
    } else {
        SIGNATURE_INTEGRITY_STATUS st = [self.sigObj checkIntegrity:hash];
        
        switch (st) {
            case INTEGRITY_MISSING:
                return(@"No integrity data found");
            case INTEGRITY_FAIL:
                return(@"Integrity failed");
            case INTEGRITY_WRONG_TYPE:
                return(@"Integrity is of wrong type");
            case INTEGRITY_OK:
                return(@"Integrity OK");
            case INTEGRITY_INVALID_KEY:
                return (@"Invalid MAC key");
        }
    }
}

- (NSString *)validationStringForDocumentData {
    SIGNATURE_HASH_TYPE hash = [self getDocHashTypeForSigObj];

    if(hash == NO_HASH) {
        return(@"No Hash Data");
    } else {
        SIGNATURE_SIGNED_DATA_STATUS st = [self.sigObj checkSignatureData:[self getDocHashTypeForSigObj] withData:self.hashData];
        
        switch (st) {
            case DATA_BAD_HASH:
                return(@"Data hash does not match. Document differs.");
            case DATA_BAD_TYPE:
                return(@"Incorrect hash type");
            case DATA_NO_HASH:
                return(@"No document hash data defined");
            case DATA_NO_DATA:
                return(@"No hash data provided");
            case DATA_GOOD:
                return(@"Data hash matches. Document is the same.");
        }
    }
}

- (IBAction)extraItemsButtonTapped:(id)sender {
    UITableViewController *vc = [[UITableViewController alloc] initWithStyle:UITableViewStylePlain];
    vc.tableView.delegate = self;
    vc.tableView.dataSource = self;
    vc.tableView.rowHeight = 33.0f;
    vc.preferredContentSize = CGSizeMake(500.0f, 33.0f * self.sigObj.extraDataItems.count);
    self.extraDatapopoverController = [[UIPopoverController alloc] initWithContentViewController:vc];

    [self.extraDatapopoverController presentPopoverFromRect:self.extraItemsButton.frame inView:self.integView permittedArrowDirections:UIPopoverArrowDirectionAny animated:YES];
}

#pragma mark - UIPopover Delegate methods

- (void)popoverController:(UIPopoverController *)popoverController willRepositionPopoverToRect:(inout CGRect *)rect inView:(inout UIView **)view {
    [popoverController dismissPopoverAnimated:YES];
    self.extraDatapopoverController = nil;
}

#pragma mark - UITableView delegate methods

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return(self.sigObj.extraDataItems.count);
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"extraData"];
    
    if(!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:@"extraData"];
        [cell.textLabel setFont:[UIFont fontWithName:@"Helvetica" size:16]];
        [cell.textLabel setTextColor:[UIColor colorWithRed:0.31f green:0.31f blue:0.31f alpha:1.0f]];
        [cell.detailTextLabel setFont:[UIFont fontWithName:@"Helvetica" size:16]];
        [cell.detailTextLabel setTextColor:[UIColor colorWithRed:0.5f green:0.5f blue:0.5f alpha:1.0f]];
    }
    
    NSString *key = self.sigObj.extraDataItems.allKeys[indexPath.row];
    cell.textLabel.text = key;
    cell.detailTextLabel.text = self.sigObj.extraDataItems[key];
    
    return(cell);
}
@end

File: Samples/Swift/SignatureSDK-S/SignatureSDK-S/SignatureIntegrityController.swift
//
//  SignatureIntegrityController.swift
//  SignatureSDK-S
//
//  Created by Joss Giffard-Burley on 28/08/2014.
//  Copyright (c) 2014 Wacom. All rights reserved.
//

import UIKit

class SignatureIntegrityController: UIViewController, UIPopoverControllerDelegate, UITableViewDataSource, UITableViewDelegate {
    //==================================================================================================================
    // MARK: Class vars
    //==================================================================================================================
    
    var sigObj :SignatureObject?
    var hashData :Data?
    
    @IBOutlet fileprivate var okButton :UIButton!
    @IBOutlet fileprivate var who :UILabel!
    @IBOutlet fileprivate var when :UILabel!
    @IBOutlet fileprivate var why :UILabel!
    @IBOutlet fileprivate var captureArea :UILabel!
    @IBOutlet fileprivate var signatureArea :UILabel!
    @IBOutlet fileprivate var digitizer :UILabel!
    @IBOutlet fileprivate var driver :UILabel!
    @IBOutlet fileprivate var machineOS :UILabel!
    @IBOutlet fileprivate var hashType :UILabel!
    @IBOutlet fileprivate var integrityStatus :UILabel!
    @IBOutlet fileprivate var documentHashType :UILabel!
    @IBOutlet fileprivate var documentHashStatus :UILabel!
    @IBOutlet fileprivate var extraItemsLabel :UILabel!
    @IBOutlet fileprivate var extraItemsButton :UIButton!
    @IBOutlet fileprivate var integView :UIView!
    
    fileprivate var extraDataPopoverController :UIPopoverController?
    
    //==================================================================================================================
    // MARK: UIViewController Methods
    //==================================================================================================================
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        okButton.layer.cornerRadius = 5.0
        extraItemsButton.setImage(UIImage(named: "action_about")?.tintedImageWithColour(UIColor(red: 0.5, green: 0.5, blue: 0.5, alpha: 1.0)), for: UIControl.State())
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let sig = sigObj {
            who.text = sig.who
            why.text = sig.why
            let bounds = "\(Int(sig.captureAreaBounds.width))x\(Int(sig.captureAreaBounds.height))"
            print(bounds)
            when.text = DateFormatter.localizedString(from: sig.when, dateStyle: .short, timeStyle: .short)
            captureArea.text =  "\(Int(sig.captureAreaBounds.width))x\(Int(sig.captureAreaBounds.height))"
            signatureArea.text =  "\(Int(sig.signatureAreaBounds.width))x\(Int(sig.signatureAreaBounds.height))"
            digitizer.text = sig.additionalData[kCaptureDigitizer]! as? String
            driver.text = sig.additionalData[kCaptureDigitizerDriver]! as? String
            machineOS.text = sig.additionalData[kCaptureMachineOS]! as? String
            hashType.text = SignatureIntegrityController.stringForHashType(getHashTypeForSigObj())
            integrityStatus.text = validationStringForSignature()
            documentHashType.text = SignatureIntegrityController.stringForHashType(getDocHashTypeForSigObj())
            documentHashStatus.text = validationStringForDocumentData()
            
            if sig.extraDataItems().count > 0 {
                extraItemsLabel.text = "\(sig.extraDataItems().count) Extra Data Items:"
            } else {
                extraItemsLabel.isHidden = true
                extraItemsButton.isHidden = true
            }
        }
    }
    
    //==================================================================================================================
    // MARK: Instance Methods
    //==================================================================================================================
    
    class func stringForHashType(_ hashType :SIGNATURE_HASH_TYPE) -> String {
        switch hashType {
        case .MD5_HASH:
            return "MD5"
        case .MD5_MAC_HASH:
            return "MD5 with MAC-KEY"
        case .SHA_1_HASH:
            return "SHA-1"
        case .SHA_224_HASH:
            return "SHA-224"
        case .SHA_256_HASH:
            return "SHA-256"
        case .SHA_384_HASH:
            return "SHA-384"
        case .SHA_512_HASH:
            return "SHA-512"
        case .NO_HASH:
            return "No Hash"
        }
    }
    
    func getHashTypeForSigObj() -> SIGNATURE_HASH_TYPE {
        var rv = SIGNATURE_HASH_TYPE.NO_HASH

        if let sig = sigObj {
            let status :SIGNATURE_INTEGRITY_STATUS = sig.checkIntegrity(.MD5_HASH)
            
            switch status {
            case .INTEGRITY_MISSING:
                rv = .NO_HASH
            case .INTEGRITY_OK, .INTEGRITY_FAIL, .INTEGRITY_INVALID_KEY, .INTEGRITY_WRONG_TYPE:
                rv = .MD5_HASH
            }
            
            let hashTypes :[SIGNATURE_HASH_TYPE] = [ .SHA_1_HASH, .SHA_224_HASH, .SHA_256_HASH, .SHA_384_HASH, .SHA_512_HASH, .MD5_MAC_HASH ]
          
            for shaType in hashTypes {
                if(sig.checkIntegrity(shaType) != SIGNATURE_INTEGRITY_STATUS.INTEGRITY_WRONG_TYPE) {
                    return shaType
                }
            }
            
        }
        
        return rv
    }
    
    func validationStringForSignature() -> String {
        let hash = getHashTypeForSigObj()
        
        if hash == .NO_HASH {
            return "No Hash"
        }
    
        if let sig = sigObj {
            switch sig.checkIntegrity(hash) {
            case .INTEGRITY_FAIL:
                return "Integrity failed"
            case .INTEGRITY_INVALID_KEY:
                return "Invalid MAC key"
            case .INTEGRITY_MISSING:
                return "No integrity data found"
            case .INTEGRITY_OK:
                return "Integrity OK"
            case .INTEGRITY_WRONG_TYPE:
                return "Integrity is of wrong type"
            }
        }
        return "No Hash"
    }
    
    func getDocHashTypeForSigObj() -> SIGNATURE_HASH_TYPE {
        let data = hashData
        
        if(data == nil) {
            return .NO_HASH
        }
        
        if let sig = sigObj {
            let status = sig.checkSignatureData(.MD5_HASH, with: data)
            
            if (status == SIGNATURE_SIGNED_DATA_STATUS.DATA_NO_DATA) || (status == SIGNATURE_SIGNED_DATA_STATUS.DATA_NO_HASH) {
                return .NO_HASH
            }
            
            let hashTypes :[SIGNATURE_HASH_TYPE] = [ .SHA_1_HASH, .SHA_224_HASH, .SHA_256_HASH, .SHA_384_HASH, .SHA_512_HASH, .MD5_MAC_HASH ]

            for type in hashTypes {
                if sig.checkSignatureData(type, with: data) != .DATA_BAD_TYPE {
                    return type
                }
            }
        }
        
        return SIGNATURE_HASH_TYPE.NO_HASH
    }
    
    func validationStringForDocumentData() -> String {
        let hash = getDocHashTypeForSigObj()
        
        if hash == .NO_HASH {
            return "No Hash"
        }
        
        if let sig = sigObj {
            switch sig.checkSignatureData(hash, with: hashData) {
            case .DATA_BAD_HASH:
                return "Data hash does not match. Document differs."
            case .DATA_BAD_TYPE:
                return "Incorrect hash type."
            case .DATA_GOOD:
                return "Data hash matches. Document is the same."
            case .DATA_NO_DATA:
                return "No hash data provided"
            case .DATA_NO_HASH:
                return "No document hash data defined"
            }
        }
        
        return "No Hash"
    }
    
    //==================================================================================================================
    // MARK: Action Callback Methods
    //==================================================================================================================
    
    @IBAction func extraItemsButtonTapped(_ sender: AnyObject) {
        let vc = UITableViewController(style: .plain)
        let ct :CGFloat = CGFloat(sigObj!.extraDataItems().count)
        
        vc.tableView.delegate = self
        vc.tableView.dataSource = self
        vc.tableView.rowHeight = 33.0
        vc.preferredContentSize = CGSize(width: 500.0, height: 33.0 * ct)
        
        extraDataPopoverController = UIPopoverController(contentViewController: vc)
        extraDataPopoverController?.present(from: extraItemsButton.frame, in: integView, permittedArrowDirections: .any, animated: true)
    }
    
    //==================================================================================================================
    // MARK: UIPopoverController delegate methods
    //==================================================================================================================

    func popoverController(_ popoverController: UIPopoverController, willRepositionPopoverTo rect: UnsafeMutablePointer<CGRect>, in view: AutoreleasingUnsafeMutablePointer<UIView>) {
        popoverController.dismiss(animated: true)
        extraDataPopoverController = nil
    }
    
    //==================================================================================================================
    // MARK: UITableView delegate methods
    //==================================================================================================================
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let sig = sigObj {
            return sig.extraDataItems().count
        } else {
            return 0
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCell(withIdentifier: "extradata") as UITableViewCell?
        
        if cell == nil {
            cell = UITableViewCell(style: .value1, reuseIdentifier: "extradata")
            cell?.textLabel!.font = UIFont(name: "Helvetica", size: 16)
            cell?.textLabel!.textColor = UIColor(red: 0.31, green: 0.31, blue: 0.31, alpha: 1.0)
            cell?.detailTextLabel?.font =  UIFont(name: "Helvetica", size: 16)
            cell?.detailTextLabel?.textColor = UIColor(red: 0.5, green: 0.5, blue: 0.5, alpha: 1.0)
        }
        
        if let sig = sigObj {
            var dict = sig.extraDataItems() as! [String: String]
            let key = Array(sig.extraDataItems().keys)[indexPath.row] as! String
            cell?.textLabel!.text = key
            cell?.detailTextLabel?.text = dict[key]
        }
        
        return cell!
    }
}
